src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,229 @@
+from itertools import product
+
+import numpy as np
+
+from qibo import gates, symbols
+from qibo.backends import GlobalBackend
+from qibo.hamiltonians import SymbolicHamiltonian
+from qibo.models import Circuit
+
+
+def prepare_states(k, nqubits):
+    """Prepares a quantum circuit in a particular state indexed by k.
+
+        Args:
+        k (int): The index of the state to be prepared.
+        nqubits (int): Number of qubits in the circuit
+
+    Returns:
+        circuit (:class:`qibo.models.Circuit`): Qibo circuit.
+    """
+
+    if nqubits != 1 and nqubits != 2:
+        raise ValueError(
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+        )
+
+    gates_list = [(gates.I,), (gates.X,), (gates.H,), (gates.H, gates.S)]
+    k_to_gates = list(product(gates_list, repeat=nqubits))[k]
+    circ = Circuit(nqubits, density_matrix=True)
+    for q in range(len(k_to_gates)):
+        gate_q = [gate(q) for gate in k_to_gates[q]]
+        circ.add(gate_q)
+    return circ
+
+
+def measurement_basis(j, circ):
+    """Implements a measurement basis for circuit.
+
+        Args:
+        circuit (:class:`qibo.models.Circuit`): Qibo circuit.
+        j (int): The index of the measurement basis.
+
+    Returns:
+        circuit (:class:`qibo.models.Circuit`): Qibo circuit.
+    """
+
+    nqubits = circ.nqubits
+    if nqubits != 1 and nqubits != 2:
+        raise ValueError(
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+        )
+
+    meas_list = [gates.Z, gates.X, gates.Y, gates.Z]
+    j_to_measurements = list(product(meas_list, repeat=nqubits))[j]
+    new_circ = circ.copy()
+    for q in range(len(j_to_measurements)):
+        meas_q = j_to_measurements[q]
+        new_circ.add(gates.M(q, basis=meas_q))
+
+    return new_circ
+
+
+def reset_register(circuit, invert_register):
+    """Returns an inverse circuit of the selected register.
```

Comment:
This function seems to be inverting a light cone of a qubit, which could be done by simply using the `circuit.light_cone` and `Circuit.invert` methods

Reply:
Maybe I'm understanding this wrong, but if by "gates in the register" you mean "gates that influence an marginal expectation value of a given qubit", then that's exactly what `Circuit.light_cone` returns.

Reply:
I see. Actually, thinking ahead in the context of the basis operations applied to a circuit, if we are given the following circuit,
```
q0: ─H─o──RX─
q1: ───RX─H──
```
we then implement basis operations like this
```
q0: ─H─B1─o──B2─RX─B3─
q1: ─────RX─B2─H──B4─
```
Here, ```B2``` is a two qubit basis operation while the rest are single qubit basis operations.

If ```B1``` happens to be the "prepare state |0>" circuit. we just apply another Hadamard to invert the initial Hadamard because we cannot call the ```ResetChannel``` function for hardware.

If ```B4``` happens to be "prepare state |0>" circuit, any idea how we can invert the qubit ```q1``` at ```B4```'s position to reset the qubit?

Reply:
There are many different protocols for GST (e.g. linear, long-sequence, advanced long-sequence, compressed, etc...). Which one exactly are you trying to implement here?

Reply:
For PEC, the methodology we follow is according to this paper:
[https://arxiv.org/pdf/1712.09271.pdf](https://arxiv.org/pdf/1712.09271.pdf
) in Section IX. I believe this is considered linear gate set tomography.

Reply:
The issue of resetting the qubits only arises when calling `GST` with the basis operations in PEC, right? In that case, we have to reset only one of the two qubits when considering two-qubit basis operations, which are the tensor product of two single-qubit basis operations. Therefore, the method you have implemented or Renato's suggestion should work.

``` python
if invert_register == "sp_0":
    invert_register = (0,)
elif invert_register == "sp_1":
    invert_register = (1,)
elif invert_register == "sp_t":
    invert_register = (0, 1)
reset_circuit = circ.light_cone(*invert_register)
inverted_circuit = reset_circuit.invert()
circ.add(inverted_circuit.on_qubits(*invert_register))
```

Reply:
Thanks! I'm thinking further ahead than GST in the implementation of PEC: What happens when ```B4``` demands that we reset to |0>?
```
q0: ─H─B1─o──B2─RX─B3─
q1: ─────RX─B2─H──B4─
```
Allow me some time to work out whether ```circ.light_cone(*invert_register)``` works as intended.